cpp/src/parquet/encoding_test.cc

```diff
@@ -1908,4 +1909,249 @@ TEST(DeltaLengthByteArrayEncodingAdHoc, ArrowDirectPut) {
+      values = rag.BinaryWithRepeats(kSize, kNumUnique, kMinLength, kMaxLength,
+                                     kNullProbability);
+      CheckDirectPut(values);
+    }
+  }
+
+  void CheckRoundtrip() override {
+    using ArrowType = typename EncodingTraits<Type>::ArrowType;
+    using IsFixedSizeBinary = ::arrow::is_fixed_size_binary_type<ArrowType>;
+
+    if constexpr (IsFixedSizeBinary::value) {
+      CheckRoundtripFLBA();
+    } else {
+      CheckRoundtripByteArray();
+    }
+  }
+
+ protected:
+  USING_BASE_MEMBERS();
+};
+
+TYPED_TEST_SUITE(DeltaByteArrayEncodingDirectPut, TestDeltaByteArrayEncodingTypes);
+
+TYPED_TEST(DeltaByteArrayEncodingDirectPut, DirectPut) {
+  ASSERT_NO_FATAL_FAILURE(this->CheckRoundtrip());
+}
+
+TEST(DeltaByteArrayEncodingAdHoc, ArrowDirectPut) {
+  auto CheckEncode = [](const std::shared_ptr<::arrow::Array>& values,
+                        const std::shared_ptr<Buffer>& encoded) {
+    auto encoder = MakeTypedEncoder<ByteArrayType>(Encoding::DELTA_BYTE_ARRAY);
+    ASSERT_NO_THROW(encoder->Put(*values));
+    auto buf = encoder->FlushValues();
+    ASSERT_TRUE(encoded->Equals(*buf));
+  };
+
+  auto ArrayToInt32Vector = [](const std::shared_ptr<::arrow::Array>& lengths) {
+    std::vector<int32_t> vector;
+    auto data_ptr = checked_cast<::arrow::Int32Array*>(lengths.get());
+    for (int i = 0; i < lengths->length(); ++i) {
+      vector.push_back(data_ptr->GetView(i));
+    }
+    return vector;
+  };
+
+  auto CheckDecode = [](std::shared_ptr<Buffer> buf,
+                        std::shared_ptr<::arrow::Array> values) {
+    int num_values = static_cast<int>(values->length());
+    auto decoder = MakeTypedDecoder<ByteArrayType>(Encoding::DELTA_BYTE_ARRAY);
+    decoder->SetData(num_values, buf->data(), static_cast<int>(buf->size()));
+
+    typename EncodingTraits<ByteArrayType>::Accumulator acc;
+    if (::arrow::is_string(values->type()->id())) {
+      acc.builder = std::make_unique<::arrow::StringBuilder>();
+    } else {
+      acc.builder = std::make_unique<::arrow::BinaryBuilder>();
+    }
+
+    ASSERT_EQ(num_values,
+              decoder->DecodeArrow(static_cast<int>(values->length()),
+                                   static_cast<int>(values->null_count()),
+                                   values->null_bitmap_data(), values->offset(), &acc));
+
+    std::shared_ptr<::arrow::Array> result;
+    ASSERT_OK(acc.builder->Finish(&result));
+    ASSERT_EQ(num_values, result->length());
+    ASSERT_OK(result->ValidateFull());
+
+    auto upcast_result = CastBinaryTypesHelper(result, values->type());
+    ::arrow::AssertArraysEqual(*values, *upcast_result);
+  };
+
+  auto CheckEncodeDecode =
+      [&](std::string_view values, std::shared_ptr<::arrow::Array> prefix_lengths,
+          std::shared_ptr<::arrow::Array> suffix_lengths, std::string_view suffix_data) {
+        auto encoded =
+            ::arrow::ConcatenateBuffers({DeltaEncode(ArrayToInt32Vector(prefix_lengths)),
```

Comment:
Just an optional suggestion, but you could rewrite `DeltaEncode` to something like:
```c++
std::shared_ptr<Buffer> DeltaEncode(::arrow::util::span<int32_t> lengths) {
  auto encoder = MakeTypedEncoder<Int32Type>(Encoding::DELTA_BINARY_PACKED);
  encoder->Put(lengths.data(), static_cast<int>(lengths.size()));
  return encoder->FlushValues();
}

std::shared_ptr<Buffer> DeltaEncode(const Int32Array& lengths) {
  return DeltaEncode(lengths.raw_values(), lengths.length());
}
```

and then you don't need `ArrayToInt32Vector` anymore.

(btw, `arrow::util::span` is a backport of [`std::span`](https://en.cppreference.com/w/cpp/container/span))

Reply:
Is this approximately what you had in mind? https://github.com/apache/arrow/pull/14341/commits/d808c5a8df6fa520064872760e0ec224248efe68

Reply:
Yes!